AppKit/CPWindow/CPWindow.j

```diff
@@ -3844,59 +3854,139 @@ var interpolate = function(fromValue, toValue, progress)
     if (![_trackingAreaViews containsObjectIdenticalTo:trackingAreaView])
         [_trackingAreaViews addObject:trackingAreaView];
     
-    // If CPTrackingAssumeInside option is set, put the tracking area in the _mouseEnteredStack
+    // If CPTrackingAssumeInside option is set, insert the tracking area in the events management system
+    // in order to have the first event sent only when mouse leaves the tracking area
     
-    if ([trackingArea options] & CPTrackingAssumeInside)
-        [_mouseEnteredStack addObject:trackingArea];
+    [self _insertTrackingArea:trackingArea assumeInside:([trackingArea options] & CPTrackingAssumeInside)];
 }
 
 - (void)_removeTrackingArea:(CPTrackingArea)trackingArea
 {
     // If mouse is in the tracking area, we remove it from the stack to avoid to fire a future mouseExited event
     
-    [_mouseEnteredStack removeObjectIdenticalTo:trackingArea];
+    [self _purgeTrackingArea:trackingArea];
     
     var trackingAreaView = [trackingArea view];
     
     [_trackingAreaViews removeObjectIdenticalTo:trackingAreaView];
 }
 
-- (void)_handleTrackingAreaEvent:(CPEvent)anEvent
+- (void)_insertTrackingArea:(CPTrackingArea)trackingArea assumeInside:(BOOL)assumeInside
 {
-    var mouseEnteredStack = [],
-        cursorUpdateStack = [],
-        point             = [anEvent locationInWindow],
-        dragging          = ([anEvent type] !== CPMouseMoved);
-
-    // Handle mouse entering tracking areas (and calc mouseEnteredStack and cursorUpdateStack)
-    
-    [self _handleMouseMovedAndEnteredEventsForEvent:anEvent atPoint:point dragging:dragging mouseEnteredStack:mouseEnteredStack cursorUpdateStack:cursorUpdateStack];
+    if (_handlingTrackingAreaEvent)
+        _restartHandlingTrackingAreaEvent = YES;
 
-    // Handle mouse exiting tracking areas
+    if (assumeInside)
+        if (_handlingTrackingAreaEvent)
+            [_mouseEnteredStack addObject:trackingArea];
+        else
+            [_previousMouseEnteredStack addObject:trackingArea];
+}
 
-    [self _handleMouseExitedEventsForEvent:anEvent atPoint:point dragging:dragging mouseEnteredStack:mouseEnteredStack];
+- (void)_purgeTrackingArea:(CPTrackingArea)trackingArea
+{
+    if (_handlingTrackingAreaEvent)
+    {
+        [_mouseEnteredStack removeObjectIdenticalTo:trackingArea];
 
-    // Cursor update
+        for (var i = 0; i < _queuedEvents.length;)
+        {
+            if ([_queuedEvents[i] trackingArea] === trackingArea)
+                [_queuedEvents removeObjectAtIndex:i];
+            else
+                i++;
+        }
```

Comment:
Hi @cacaodev ! Yes, indeed (and that was my first version), but I have a bunch of (good and bad) reasons to do this... For example, I have to send events in the produced order so that mouseEntered are processed first, then mouseExited and finally cursorUpdate. This way, a mouseEntered method removing the related view will be processed in a way that next cursorUpdate event will be deleted.